common/ast/expr.go

```diff
@@ -0,0 +1,692 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ast
+
+import (
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+)
+
+// ExprKind represents the expression node kind.
+type ExprKind int
+
+const (
+	UnspecifiedKind ExprKind = iota
+	LiteralKind
+	IdentKind
+	SelectKind
+	CallKind
+	ListKind
+	MapKind
+	StructKind
+	ComprehensionKind
+)
+
+// NavigateExpr converts a protobuf Expr node and a map of expression types by id to a NavigableExpr
+func NavigateExpr(expr *exprpb.Expr, typeMap map[int64]*types.Type) NavigableExpr {
+	return newNavigableExpr(nil, expr, typeMap)
+}
+
+// ExprMatcher takes a NavigableExpr in and indicates whether the value is a match.
+//
+// This function type should be use with the `Match` and `MatchList` calls.
+type ExprMatcher func(NavigableExpr) bool
+
+// ConstantValueMatcher returns an ExprMatcher which will return true if the input NavigableExpr
+// is comprised of all constant values, such as a simple literal or even list and map literal.
+func ConstantValueMatcher() ExprMatcher {
```

Comment:
Do you have a use case in mind for this?

Reply:
I was thinking of using this for constant folding in the future to sanity check whether function calls have all constant inputs using the `MatchExpr` (note, I think `MatchExpr` needs a different name, maybe `MatchSubset`?)